juju/apiclient.py

```diff
@@ -17,10 +25,250 @@
+
+
+class ServerError(Exception):
+    """Base exception class for particular server side errors.
+
+    If there is no specific error type for the error code returned from the
+    server, then this is the type of the returned error.
+
+    """
+    def __init__(self, error_code, message):
+        self.error_code = error_code
+        self.message = message
+
+    def __str__(self):
+        return "<ServerError - '{}'>".format(message)
+
+
+class NotImplementedError(ServerError):
+    """NotImplementedError is raised when a method does not exist.
+
+    The error could be caused due to either the method name being incorred, or
+    the facade version not being supported on the API server, or the facade
+    name itself not existing on the server.
+
+    If the Facade instances are being constructed through the Connection
+    instance, then the more specialized UnknownFacade and
+    FacadeVersionNotSupported errors will be raised, however if a Facade
+    instance is created through another method, the NotImplementedError will
+    be raised in the cases of bad facade names or versions.
+
+    """
+    def __init__(self, *args):
+        super(NotImplementedError, self).__init__(*args)
+
+
+
+def new_error(response):
+    """Constructs a specific exception type based on the ErrorCode."""
+    err_code = response.get("ErrorCode")
+    message = response.get("Error")
+    if err_code == "not implemented":
+        return NotImplementedError(err_code, message)
+    return ServerError(err_code, message)
+
+
+class Facade(object):
+    """The Facade instances provide convenient syntactic sugar for API calls.
+
+    Facade instances are constructed with a Connection, and facade name and
+    version. Any resulting method calls to a facade are translated into RPC
+    calls on the connection instance.
+
+    For example, calling the 'FullStatus' method on the client facade:
+
+        >>> client = Facade(connection, "Client")
+        >>> result = client.FullStatus()
+
+    This ends up being the same as:
+
+        >>> connection._rpc("Client", "FullStatus")
+
+    """
+
+    def __init__(self, connection, name, version=None):
+        self.connection = connection
+        self.name = name
+        self.version = version
+
+    def __getattribute__(self, attr):
+        try:
+            return object.__getattribute__(self, attr)
+        except AttributeError:
+            return functools.partial(self.connection._rpc, self.name, attr, version=self.version)
+
+
+class Connection(object):
+    """A connection represents an open, authenticated API connection."""
+    _request_id = 0
+    _upgrade_retry_count = 60
+    _upgrade_retry_delay_secs = 1
+
+    def __init__(self, address, cacert, auth_tag, password, nonce="", env_uuid=""):
+        self._cert_file = self._write_cert(cacert)
+        endpoint = self._endpoint(address, env_uuid)
+        cert_path = self._cert_file.name
+        self._connection = self._connect(endpoint, cert_path)
```

Comment:
No, not at this stage. As the object is garbage collected it will close the websocket.

This way we never have to actually look to see if the connection is open first.

We could add one later, along with checks for the connection being open as it becomes necessary.